modules/Router.js

```diff
@@ -54,6 +59,27 @@ const Router = React.createClass({
   },
 
   componentWillMount() {
+    const { history } = this.props
+    if (isDeprecatedHistory(history)) {
+      this.setupDeprecatedHistory()
+    } else {
+      this.setupTransitionManager()
+    }
+  },
+
+  setupTransitionManager() {
+    const { history, routes, children } = this.props
+    this.transitionManager = createTransitionManager(history, createRoutes(routes || children))
+    this.transitionManager.listen((error, state) => {
+      if (error) {
+        this.handleError(error)
+      } else {
+        this.setState(state, this.props.onUpdate)
+      }
+    })
+  },
```

Comment:
If you build the entire router object here, then you don't need to split it out. I'm thinking something like:

```js
const { addRouteLeaveHook, isActive } = transitionManager
this.router = {
  ...history,
  addRouteLeaveHook,
  isActive,
  get location() {
    return this.state.location
  }
}
```

The `location` getter is a little bit messy though.